src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -382,37 +384,30 @@ public static byte[] UrlEncodeToBytes(byte[] value, int offset, int count)
                 return null;
             }
 
-            int cSpaces = 0;
-            int cUnsafe = 0;
+            bool foundSpaces = false;
+            int unsafeCount = 0;
 
             // count them first
             for (int i = 0; i < count; i++)
             {
                 char ch = (char)value[offset + i];
 
                 if (ch == ' ')
-                    cSpaces++;
+                    foundSpaces = true;
                 else if (!IsUrlSafeChar(ch))
-                    cUnsafe++;
+                    unsafeCount++;
             }
 
             // nothing to expand?
-            if (cSpaces == 0 && cUnsafe == 0)
+            if (!foundSpaces && unsafeCount == 0)
```

Comment:
@hughbe No, I meant something like this:

```cs
if (unsafeCount == 0)
{
    var subarray = new byte[count];
    if (!foundSpaces)
    {
        Buffer.BlockCopy(value, offset, subarray, 0, count);
    }
    else
    {
        for (int i = 0; i < count; i++) 
        {
            if (value[i] == (byte)' ')
            {
                subarray[i] = (byte)'+';
            }
            else
            {
                subarray[i] = value[i];
            }
        }
    }
    return subarray;
}
```

Since `Buffer.BlockCopy` will vary in time complexity depending on the length of the array, as will making a pass over it via the for-loop.